### PR TITLE
Implement Service to Configure Mail Settings

### LINKED ISSUE #1295: Implement Service to Configure Mail Settings
**Actual Behaviour**

Currently, static values are fed

**Expected Behaviour**

For Issue https://github.com/fossasia/susi.ai/issues/2454, Implement Service to Configure Mail Settings

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by fragm3**
Fixes #1295

Changes: The SUSI.AI server has email settings build in currently. Implement Service to Configure Mail Settings.

### CODE REVIEW

**Comment by fragm3** [comment]
For settings mail settings, API point: http://localhost:4000/aaa/emailSettings.json?access_token=tO063rvh13K58CsZPywwSu4qJqjUyL&frontend_url=https://susi.ai&displayname=shubham&type=SMTP&name=Admin&email=server@loklak.org&sendgrid_token=dfg

Getting response:
```
{
  "session": {"identity": {
    "type": "host",
    "name": "0:0:0:0:0:0:0:1_ba69b956",
    "anonymous": true
  }},
  "accepted": true,
  "Successfully updated mail settings": true
}
```

Fetching mail settings, API point: http://localhost:4000/aaa/getEmailSettings.json
Getting response:

```
{
  "settings": {
    "frontend_url": "https://susi.ai",
    "name": "Admin",
    "sendgrid_token": "",
    "type": "SMTP",
    "email": "server@loklak.org"
  },
  "session": {"identity": {
    "type": "host",
    "name": "0:0:0:0:0:0:0:1_ba69b956",
    "anonymous": true
  }},
  "accepted": true,
  "message": "Success : Fetched mail settings!"
}
```

**Comment by fragm3** [comment]
@akshatnitd @Orbiter @norbusan, please review

File: conf/config.properties [hunk 1]
```diff
@@ -21,7 +21,7 @@ users.password.regex.tooltip=Enter a combination of atleast six characters
 users.admin.localonly=false
 
 # SMTP emailing switch (for password recovery and signup verification etc)
-smtp.mails.enabled=true
```
Comment (akshatnitd) [inline conf/config.properties]: Don't remove an existing field

File: src/ai/susi/server/api/aaa/EmailSettingsService.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+ * Optional - For sending a testing email, add parameter, receiver_email
+ */
+
+public class EmailSettingsService extends AbstractAPIHandler implements APIHandler {
+
+    private static final long serialVersionUID = 5000658108778105134L;
+
+    @Override
+    public UserRole getMinimalUserRole() {
+        return UserRole.ANONYMOUS;
```
Comment (akshatnitd) [inline src/ai/susi/server/api/aaa/EmailSettingsService.java]: ADMIN?

File: src/ai/susi/server/api/aaa/GetEmailSettings.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+import ai.susi.server.APIException;
+import ai.susi.server.APIHandler;
+import ai.susi.server.AbstractAPIHandler;
+import ai.susi.server.Query;
+import ai.susi.server.ServiceResponse;
+import ai.susi.server.UserRole;
+
+/**
+ * This Servlet gives a API Endpoint to fetch Skill Slideshow used by SUSI
```
Comment (akshatnitd) [inline src/ai/susi/server/api/aaa/GetEmailSettings.java]: Change description

File: src/ai/susi/server/api/aaa/GetEmailSettings.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    private static final long serialVersionUID = 5000658108778105134L;
+
+    @Override
+    public String getAPIPath() {
+        return "/aaa/getEmailSettings.json";
+    }
+
+    @Override
+    public UserRole getMinimalUserRole() {
+        return UserRole.ANONYMOUS;
```
Comment (akshatnitd) [inline src/ai/susi/server/api/aaa/GetEmailSettings.java]: ADMIN

File: conf/config.properties [hunk 5]
```diff
@@ -31,10 +31,12 @@ smtp.host.encryption=tls
 smtp.host.port=465
 
 # SMTP sender displayname (the name displayed to the recipient), email, username (to login) and password. Usually email and username are the same (but not necessarily)
-smtp.sender.displayname=Admin
```
Comment (akshatnitd) [inline conf/config.properties]: @Orbiter Need suggestion, if these config key name should be changed.

File: conf/config.properties (hunk 1 above)
Comment (fragm3) [reply conf/config.properties]: Actually, this is an implication, I think, we don't require it

**Comment by fragm3** [comment]
@akshatnitd made the changes, except config keys name. Waiting for @Orbiter replies to take further changes

File: conf/config.properties (hunk 5 above)
Comment (mariobehling) [reply conf/config.properties]: Yes, the display name of the sender should be an option, that can be defined in the frontend.

File: conf/config.properties [hunk 6]
```diff
@@ -21,20 +21,19 @@ users.password.regex.tooltip=Enter a combination of atleast six characters
...
 
 # SMTP connection encryption and corresponding port (none/25, starttls/587 or tls/465). in unusual cases, different ports may be used.
 smtp.host.encryption=tls
 smtp.host.port=465
 
 # SMTP sender displayname (the name displayed to the recipient), email, username (to login) and password. Usually email and username are the same (but not necessarily)
-smtp.sender.displayname=Admin
-smtp.sender.email=server@loklak.org
+mail.displayname=Admin
+mail.email=smtp.gmail.com
```
Comment (mariobehling) [inline conf/config.properties]: where does this come from? This is not an email! Do you mean to create smtp gateways? Please check how eventyay does it.

File: src/ai/susi/EmailHandler.java [hunk 7]
```diff
@@ -41,20 +41,20 @@
      * @throws Exception on errors
      */
 	public static void sendEmail(@Nonnull String addressTo, @Nonnull String subject, @Nonnull String text) throws Exception {
-		String senderEmail = DAO.getConfig("smtp.sender.email", null);
-        String displayname = DAO.getConfig("smtp.sender.displayname", null);
+		String senderEmail = DAO.getConfig("mail.email", null);
```
Comment (mariobehling) [inline src/ai/susi/EmailHandler.java]: Should not this still be the sender email?

File: src/ai/susi/server/api/aaa/EmailSettingsService.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
+/**
+ *  APIKeysService
+ *  Copyright by Shubham Gupta @fragm3
```
Comment (mariobehling) [inline src/ai/susi/server/api/aaa/EmailSettingsService.java]: This is just following the standard implementation of sendgrid, right? I don't think it reaches the threshold for Copyright and requires an additional copyright note.

File: src/ai/susi/server/api/aaa/GetEmailSettings.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
+/**
+ *  GetApiKeys
+ *  Copyright by Shubham Gupta @fragm3
+ *
```
Comment (mariobehling) [inline src/ai/susi/server/api/aaa/GetEmailSettings.java]: This is too basic for adding additional copyright notes. Some people started in the past to add copyright notices. It does not make sense that every single file gets its own copyright now. The whole project already has a copyright.

File: conf/config.properties (hunk 6 above)
Comment (fragm3) [reply conf/config.properties]: Okay, will look into it

**Comment by fragm3** [comment]
@akshatnitd @mariobehling made the config keys same

File: src/ai/susi/server/api/aaa/EmailSettingsService.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+                        "You have successfully sent a email");
+                result.put("accepted", true);
+                result.put("message", "You have successfully sent a email to your email id:" + adminEmail);
+            } catch (Exception e) {
+                result.put("message", e.getMessage() + receiverEmail);
+            }
+            return new ServiceResponse(result);
+        }
+
+        DAO.setConfig("smtp.sender.email", email);
```
Comment (akshatnitd) [inline src/ai/susi/server/api/aaa/EmailSettingsService.java]: Add defaults for this config values on the config file

Comment (fragm3) [reply src/ai/susi/server/api/aaa/EmailSettingsService.java]: Already added on line 34 in `config.properties`

Comment (fragm3) [reply src/ai/susi/server/api/aaa/EmailSettingsService.java]: Edit: Reset all configs, please review

**Pull request merged**